File: authentication-client/authentication-client-java/src/main/java/co/cask/cdap/security/authentication/client/basic/BasicAuthenticationClient.java
/*
 * Copyright 2014 Cask Data, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package co.cask.cdap.security.authentication.client.basic;

import co.cask.cdap.security.authentication.client.AbstractAuthenticationClient;
import co.cask.cdap.security.authentication.client.AccessToken;
import co.cask.cdap.security.authentication.client.Credential;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import org.apache.commons.codec.binary.Base64;
import org.apache.commons.lang.StringUtils;
import org.apache.http.HttpHeaders;
import org.apache.http.client.methods.HttpGet;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.List;
import java.util.Properties;

/**
 * Authentication client that supports "Basic access authentication" using username and password.
 */
public class BasicAuthenticationClient extends AbstractAuthenticationClient {
  private static final Logger LOG = LoggerFactory.getLogger(BasicAuthenticationClient.class);

  private static final String AUTHENTICATION_HEADER_PREFIX_BASIC = "Basic ";
  private static final String USERNAME_PROP_NAME = "security.auth.client.username";
  private static final String PASSWORD_PROP_NAME = "security.auth.client.password";

  private String username;
  private String password;
  private final List<Credential> credentials;

  /**
   * Constructs new instance.
   */
  public BasicAuthenticationClient() {
    super();
    credentials = ImmutableList.of(new Credential(USERNAME_PROP_NAME, "Username for basic authentication.", false),
                                   new Credential(PASSWORD_PROP_NAME, "Password for basic authentication.", true));
  }

  @Override
  public void configure(Properties properties) {
    if (StringUtils.isNotEmpty(username) || StringUtils.isNotEmpty(password)) {
      throw new IllegalStateException("Client is already configured!");
    }

    username = properties.getProperty(USERNAME_PROP_NAME);
    Preconditions.checkArgument(StringUtils.isNotEmpty(username), "The username property cannot be empty.");

    password = properties.getProperty(PASSWORD_PROP_NAME);
    Preconditions.checkArgument(StringUtils.isNotEmpty(password), "The password property cannot be empty.");

    LOG.debug("Basic authentication client is configured successfully.");
  }

  @Override
  public List<Credential> getRequiredCredentials() {
    return credentials;
  }

  @Override
  protected AccessToken fetchAccessToken() throws IOException {
    if (StringUtils.isEmpty(username) || StringUtils.isEmpty(password)) {
      throw new IllegalStateException("Base authentication client is not configured!");
    }

    LOG.debug("Authentication is enabled in the gateway server. Authentication URI {}.", getAuthUrl());
    HttpGet getRequest = new HttpGet(getAuthUrl());

    String auth = Base64.encodeBase64String(String.format("%s:%s", username, password).getBytes());
    auth = auth.replaceAll("(\r|\n)", StringUtils.EMPTY);
    getRequest.addHeader(HttpHeaders.AUTHORIZATION, AUTHENTICATION_HEADER_PREFIX_BASIC + auth);

    return execute(getRequest);
  }
}

File: authentication-client-python/BasicAuthenticationClient.py
import base64
import logging
import urllib2
from flask import json
from AbstractAuthenticationClient import AbstractAuthenticationClient
from Config import Config
from Credential import Credential

LOG = logging.getLogger(__name__)


class BasicAuthenticationClient(AbstractAuthenticationClient):

    USERNAME_PROP_NAME = u'security.auth.client.username'
    PASSWORD_PROP_NAME = u'security.auth.client.password'

    def __init__(self):
        super(BasicAuthenticationClient, self).__init__()
        self.__username = None
        self.__password = None
        self.__credentials = (Credential(self.USERNAME_PROP_NAME, u'Username for basic authentication.', False),
                              Credential(self.PASSWORD_PROP_NAME, u'Password for basic authentication.', True))

    def get_required_credentials(self):
        return self.__credentials

    def fetch_access_token(self):
        if not self.__username or not self.__password:
            raise ValueError(u'Base authentication client is not configured!')
        LOG.debug(u'Authentication is enabled in the gateway server. Authentication URI {}.', self.auth_url)

        base64string = base64.encodestring('%s:%s' % (self.__username, self.__password)).replace('\n', '')
        auth_header = json.dumps( {u"Authorization": u"Basic %s" % base64string })

        return self.execute(auth_header)

    def configure(self, config_file):
        if self.__username or self.__password:
            raise ValueError(u'Client is already configured!')
        config = Config().read_from_file(config_file)
        self.__username = config.username
        if not self.__username:
            raise ValueError(u'The username property cannot be empty.')

        self.__password = config.password
        if not self.__password:
            raise ValueError(u'The password property cannot be empty.')
